### ISSUE TITLE
Potential Ablation Studies on Scaling Law in EAGLE3

### ISSUE THREAD

**Issue by luyuzhe111** [SafeAILab/EAGLE#220]
Dear EAGLE authors,

Congrats on a great series of work! I have a suggestion on performing an additional ablation study to understand the impact of architectural changes, in particular, adding proper RMS norms to EAGLE 3 on the emergence of scaling law in the EALGE 3 paper. 

If I understand correctly, there are a couple architectural changes aside from the multi-layer feature fusion, listed as below.

- The use of last hidden states prior to the final output norm [[line](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L687)]. Since there is no hidden states collection script yet, I had to take a guess here. 
- The addition of hidden states norm & embedding norm prior to the FC layer [[line1](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L522), [line2](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L395)]. 
- The addition of a final output norm [[line](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L522)]. 

Essentially, the DeepSeek MTP design with a missing pre-attention norm as I mentioned in https://github.com/SafeAILab/EAGLE/issues/219. 

I believe these normalization layers play a key role in the emergence of scaling law mentioned in the paper, in addition to the loss function change. I hypothesize this way because of the following empirical observation:

previously, the last hidden states, **after** the [final output norm](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L582) are used as inputs to EAGLE.  This has been making EAGLE training unstable, and potentially unscalable, because those hidden states somehow have very large norms, which in turn, forced the weights of the projection (`fc`) layer to be quite small. Through analysis, one can find that there is something special about this final output norm. Somehow, its scaling factors are multitudes large than all previous RMS norms in terms of magnitude, leading to hidden states with very large norms. By using un-normalized last hidden states or hidden state RMS norm before the `fc` layer, this difficulty in training will simply go away and one can easily use much larger learning rates and get better results even without more data.

Given the EAGLE series of work has gathered a lot of interest in both academia and industry, I would really appreciate the authors if they could provide an additional ablation study on the impact of these architectural changes, so as to help the research community better understand the key factor behind EAGLE3's success. Thanks!

**Comment by carlbunny** [comment]
@luyuzhe111 I have a eagle3 training repro based on eagle1 stack without the norm changes you mentioned. So far comparing to eagle1, I have not get better TPC on short drafting length (draft length on 3 or 4). Do you feel these norms change will make big difference?

**Comment by luyuzhe111** [comment]
@carlbunny the EAGLE authors can comment best but training-time test should be pretty effective even without architectural changes. if you are not seeing better acceptance length (assume that's what you meant by TPC), then probably something else is off.

**Comment by carlbunny** [comment]
Just a self-update that I tried with and without the addition of hidden states norm & embedding norm prior to the FC layer [[line1](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L522), [line2](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py#L395)], it can increase the accurate by 5% :)
